### ISSUE TITLE
Might be worth checking out delta-dotnet for native Parquet flushing without reinventing the Delta Transaction protocol

### ISSUE THREAD

**Issue by mdrakiburrahman** [aloneguid/delta-dotnet#2]
Hey @aloneguid,

Big fan of your parquet implementation, thanks for maintaining that.

FYI, we've been working on a delta-dotnet implementation, [here](https://github.com/delta-incubator/delta-dotnet) (I understand you've gone through it, perhaps before the Kernel support was recently added).

Currently, it passes Arrow to `delta-rs` which writes parquet, but passing large data over the FFI is memory intensive. We have an issue here to use a native C# Parquet implementation (i.e. your perhaps your parquet library), to bypass this problem:

https://github.com/delta-incubator/delta-dotnet/issues/103

The problem with Delta isn't the parquet flush, but rather the Transaction protocol. For example, if you were to try to implement [Deletion vector support](https://delta.io/blog/2023-07-05-deletion-vectors/) or [non-trivial Change Query support](https://jack-vanlightly.com/analyses/2024/9/24/change-query-support-in-delta-lake) in this new repo in C#, you'd probably have to spend a lot of time and code looking into the Java/Scala/Rust implementations, and may end up introducing subtle bugs when reinventing the wheel on that front. Once you go down all the features Delta supports, you'll find that the [Delta Protocol README](https://github.com/delta-io/delta/blob/master/PROTOCOL.md) doesn't have majority of them documented, you'll only find out how it works when reading [the Spark implementation](https://github.com/delta-io/delta/tree/master/spark/src/main) (me and my teammates had to do this multiple times).

Delta Lake authors are solving this problem once and for all using [this Kernel project](https://delta.io/blog/delta-kernel/), so Delta stops being tied to Apache Spark and wins over the community support by driving major adoption in all languages, without each language rewriting the core transaction algorithms.

Essentially, the idea is, the heavy duty work of flushing parquet can be done in some language, but the tricky "domain knowledge" part where you have to update/intelligently read the Delta Transaction log JSON files, you leave that to [the kernel](https://github.com/delta-io/delta-kernel-rs). E.g. this is [how DuckDB](https://duckdb.org/2024/06/10/delta.html) works with Delta, and it's blazing fast.

I wanted to open this issue so you knew about the vision of the `delta-dotnet` repo - since the README or the current `delta-rs` wrapper doesn't really make that obvious. 

There's 5 different Delta Dotnet/C# implementations I know of:

1. I personally also went down the route of trying to implement Delta in pure C# as well (private repo that I deleted when I found `delta-dotnet`)
2.  So did [this other repo from a solo engineer](https://github.com/johnsusi/delta-net) (which I see you mentioned [in this blog](https://www.aloneguid.uk/posts/2024/11/deltalake-dotnet/)). 
3. A closed source repo that is maintained by a team of engineers that only implements the very primitive [V1 protocol](https://docs.delta.io/2.0.0/versioning.html), since protocol upgrades are hard since there are too many Delta features to implement
4. `delta-dotnet` started wrapping `delta-kernel-rs` before now using the kernel FFI
5. Your new repo here

IMO, it'd be better if all these C# engineers worked together on one implementation that was future facing and performant, while offloading the domain knowledge of Delta Transaction protocol to the folks designing the table format.

I agree with what you said in your blog, the `delta-rs` wrap is unnatural just like the C++ wrapper Nuget on Parquet in [ParquetSharp](https://github.com/G-Research/ParquetSharp) (Memory leaks etc), and we plan on removing that delta-rs wrapper ASAP - see [discussion here](https://github.com/delta-incubator/delta-dotnet/discussions/79). 

So that we use C# to flush parquet natively, while using the Delta Kernel for the lightweight (memory wise) but complex job of transaction log updates.

I actually used your Parquet Nuget on a private branch of mine and got ridiculous flush rates (27 GB/min) from native C# while bulk updating the Delta Trx log for blind APPENDs: https://www.rakirahman.me/delta-dotnet/

So I know this method (native parquet flush, FFI trx log update) will scale.

In summary,

We'd love to align efforts over at delta-dotnet to build a fast, performant, memory-safe/native C# parquet reader/flusher, while using the Delta-kernel to offload all the "Delta Transaction Log stuff". We'd also love your help given all the great work you've done on Parquet.

---

I hope you understand where I'm coming from and am not appearing as nosy 🙂 - I want to reach out to save someone else the effort of going down the Delta Transaction protocol rabbit hole.

That being said, reading this project's README, I understand there's something pure about a no FFI implementation if it's maintained well, it's just a lot of work to keep feature parity with the evolving Delta protocol.

**Comment by mdrakiburrahman** [comment]
Either way, thanks again for the great Parquet package 🙂
